src/index.ts

```diff
@@ -476,138 +661,234 @@ export class Upload extends Pumpify {
+        'Content-Length': expectedUploadSize,
+        'Content-Range': `bytes ${this.offset}-${endingByte}/${this.contentLength}`,
+      };
+    } else {
+      headers = {
+        'Content-Range': `bytes ${this.offset}-*/${this.contentLength}`,
+      };
+    }
+
     const reqOpts: GaxiosOptions = {
       method: 'PUT',
       url: this.uri,
-      headers: {
-        'Content-Range': 'bytes ' + this.offset + '-*/' + this.contentLength,
-      },
-      body: requestStreamEmbeddedStream,
+      headers,
+      body: requestStream,
     };
 
     try {
       await this.makeRequestStream(reqOpts);
-    } catch (e) {
+    } catch (err) {
+      const e = err as Error;
       this.destroy(e);
     }
   }
 
-  private onChunk(
-    chunk: string,
-    enc: BufferEncoding,
-    next: (err?: Error, data?: string) => void
-  ) {
-    const offset = this.offset!;
-    const numBytesWritten = this.numBytesWritten;
+  // Process the API response to look for errors that came in
+  // the response body.
+  private responseHandler(resp: GaxiosResponse) {
+    if (resp.data.error) {
+      this.destroy(resp.data.error);
+      return;
+    }
 
-    this.emit('progress', {
-      bytesWritten: this.numBytesWritten,
-      contentLength: this.contentLength,
-    });
+    const shouldContinueWithNextMultiChunkRequest =
+      this.chunkSize &&
+      resp.status === RESUMABLE_INCOMPLETE_STATUS_CODE &&
+      resp.headers.range;
+
+    if (shouldContinueWithNextMultiChunkRequest) {
+      // Use the upper value in this header to determine where to start the next chunk.
+      // We should not assume that the server received all bytes sent in the request.
+      // https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+      const range: string = resp.headers.range;
+      this.offset = Number(range.split('-')[1]) + 1;
+
+      // We should not assume that the server received all bytes sent in the request.
+      // - https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+      const missingBytes = this.numBytesWritten - this.offset;
+      if (missingBytes) {
+        const dataToPrependForResending = this.lastChunkSent.slice(
+          -missingBytes
+        );
+        // As multi-chunk uploads send one chunk per request and pulls one
+        // chunk into the pipeline, prepending the missing bytes back should
+        // be fine for the next request.
+        this.unshiftChunkBuffer(dataToPrependForResending);
+        this.numBytesWritten -= missingBytes;
+        this.lastChunkSent = Buffer.alloc(0);
+      }
 
-    // check if this is the same content uploaded previously. this caches a
-    // slice of the first chunk, then compares it with the first byte of
-    // incoming data
-    if (numBytesWritten === 0) {
-      let cachedFirstChunk = this.get('firstChunk');
-      const firstChunk = chunk.slice(0, 16).valueOf();
+      // continue uploading next chunk
+      this.continueUploading();
```

Comment:
It looks like the logic of `continueUploading` hasn't changed much. Is the upload endpoint a constant URI, and it's just `Content-Range` that changes to support multiple chunk uploads?

Reply:
Correct - it's mainly `Content-Range` and supporting N requests per `Upload` (required a lot of decoupling throughout the class)